src/network/network_crypto.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file network_admin.cpp Server part of the admin network protocol. */
+
+#include "../stdafx.h"
+
+#include "network_crypto.h"
+#include "core/packet.h"
+#include "../core/random_func.hpp"
+#include "../3rdparty/monocypher/monocypher.h"
+#include "../settings_type.h"
+#include "../string_func.h"
+
+#include "../safeguards.h"
+
+/**
+ * Call \c crypto_wipe for all the data in the given span.
+ * @param span The span to cryptographically wipe.
+ */
+static void crypto_wipe(std::span<uint8_t> span)
+{
+	crypto_wipe(span.data(), span.size());
+}
+
+static ssize_t TransferOutSpan(std::span<uint8_t> destination, const char *source, size_t amount)
+{
+	size_t to_copy = std::min(amount, destination.size());
+	std::copy(source, source + to_copy, destination.data());
+	return to_copy;
+}
+
+template <size_t size>
+class SharedKeys {
+	std::array<uint8_t, 2 * size> keys;
+public:
+	~SharedKeys() { crypto_wipe(keys); }
+	constexpr std::span<const uint8_t> ClientToServer() const { return std::span(this->keys.data(), size); }
+	constexpr std::span<const uint8_t> ServerToClient() const { return std::span(this->keys.data() + size, size); }
+	constexpr std::span<uint8_t> Both() { return this->keys; }
+};
+
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+static bool X25519KeyExchange(SharedKeys<32> &shared_keys, const X25519Key &my_private_key, const X25519Key &their_public_key)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), my_private_key.data(), their_public_key.data());
```

Comment:
Btw, for your own sanity, I would use a bit of different naming here:

Monocypher calls it "secret key" and "public key". I would adapt that. The reason is, otherwise they are both `pk`, which often confuses people. But using `sk` vs `pk`, you resolve that issue.

And not doing a code-review, just functional looking into what you are up to, but `my_` and `their_` .. we need to find a better name :D The static `my_client` we have is also weird :P But okay, nitpicks is for another day.

Reply:
I think I'm going with:
- our secret key
- our public key
- peer public key
- shared secret
- derived keys

It's a bit of a mix-match from all, but probably the most different names that I could come up with.